mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -111,13 +134,58 @@ impl Bio {
                 debug!("BIO for nexus marked completed but has outstanding")
             }
         }
-
-        unsafe { spdk_bdev_io_complete(self.0, io_status::SUCCESS) };
+        self.drop_retries();
+        unsafe {
+            spdk_bdev_io_complete(self.0, io_status::SUCCESS);
+        }
     }
     /// mark the IO as failed
     #[inline]
     pub(crate) fn fail(&mut self) {
-        unsafe { spdk_bdev_io_complete(self.0, io_status::FAILED) };
+        self.drop_retries();
+        unsafe {
+            spdk_bdev_io_complete(self.0, io_status::FAILED);
+        }
+    }
+
+    /// ensure the retry_counts box is freed
+    #[inline]
+    pub(crate) fn drop_retries(&mut self) {
```

Comment:
IO's are always put back on the queue they were borrowed from, and they are fixed in size, and don't require any (de)allocation in the "hot path". I think we should try *hard* to keep it that way. Moreover, this can lead to leaks and we are throwing away, the benefits of rust in this case.

Reply:
If it's OK to retry all child IOs when one fails that would simplify things a lot and the allocation could be removed. I was trying to limit it to just the failed child. I'll make the change.